lib/llhttp/parser.js

```diff
@@ -149,127 +130,89 @@ class HTTPParser {
     return 0
   }
 
-  [kOnExecute] (ret) {}
+  [kOnUpgrade] (head) {}
 
   close () {
-    inst.exports.llhttp_free(this[kPtr])
-    this[kPtr] = null
-    inst.exports.free(this[kBufferPtr])
-    this[kBufferPtr] = null
-    this[kCurrentBuffer] = Buffer.alloc(0)
+    if (this[kPtr]) {
+      inst.exports.llhttp_free(this[kPtr])
+      this[kPtr] = null
+    }
+
+    if (this[kBufferPtr]) {
+      inst.exports.free(this[kBufferPtr])
+      this[kBufferPtr] = null
+    }
   }
 
   execute (data) {
-    currentParser = this
-    this[kCurrentBuffer] = data
+    assert(this[kPtr])
+
     // Be sure the parser buffer can contain `data`
     if (data.length > this[kBufferSize]) {
-      this[kMallocBuffer](Math.ceil(data.length / 4096) * 4096)
+      if (this[kBufferPtr]) {
+        inst.exports.free(this[kBufferPtr])
+      }
+      this[kBufferSize] = Math.ceil(data.length / 4096) * 4096
+      this[kBufferPtr] = inst.exports.malloc(this[kBufferSize])
+      // Instantiate a Unit8 Buffer view of the wasm memory that starts from the parser buffer pointer.
+      this[kBufferView] = new Uint8Array(inst.exports.memory.buffer, this[kBufferPtr], this[kBufferSize])
     }
-    // Instantiate a Unit8 Buffer view of the wasm memory that starts from the parser buffer pointer
-    // and has the size of `data`
-    const u8 = new Uint8Array(inst.exports.memory.buffer, this[kBufferPtr], data.length)
-    // Fill the view with `data`
-    u8.set(data)
+
+    this[kBufferView].set(data)
+
     // Call `execute` on the wasm parser.
     // We pass the `llhttp_parser` pointer address, the pointer address of buffer view data,
     // and finally the length of bytes to parse.
     // The return value is an error code or `constants.ERROR.OK`.
     // See https://github.com/dnlup/llhttp/blob/undici_wasm/src/native/api.c#L106
+    currentParser = this
     const err = inst.exports.llhttp_execute(this[kPtr], this[kBufferPtr], data.length)
-    let ret = data.length
-    if (err !== constants.ERROR.OK) {
-      const errorPos = inst.exports.llhttp_get_error_pos(this[kPtr])
-      ret = errorPos - this[kBufferPtr]
-      if (err === constants.ERROR.PAUSED_UPGRADE) {
-        inst.exports.llhttp_resume_after_upgrade(this[kPtr])
-      } else {
-        ret = this[kMakeError](err)
-      }
-    }
-    this[kOnExecute](ret)
     currentParser = null
-    return ret
-  }
 
-  getCurrentBuffer () {
-    return this[kCurrentBuffer]
-  }
-
-  [kMakeError] (number) {
-    const ptr = inst.exports.llhttp_get_error_reason(this[kPtr])
-    const u8 = new Uint8Array(inst.exports.memory.buffer)
-    const len = u8.indexOf(0, ptr) - ptr
-    const message = cstr(ptr, len)
-    const code = constants.ERROR[number]
-    return new HTTPParserError(message, code)
+    if (err === constants.ERROR.PAUSED_UPGRADE) {
+      const offset = inst.exports.llhttp_get_error_pos(this[kPtr]) - this[kBufferPtr]
+      this[kOnUpgrade](data.slice(offset))
+      this.close()
```

Comment:
Ah, wasn't the parser getting closed before? Could that be the cause of #659 ?

Reply:
I don’t think so. It’s just closed earlier now.